### PR TITLE
Google Classroom reauthorization flow with CSRF protection

### PR DESCRIPTION
**Pull request by islemaster**
Reverses https://github.com/code-dot-org/code-dot-org/pull/35734 and updates the Google Classroom reauthorization flow to be compatible with CSRF-protected OAuth.

image

You can think of this PR as containing four changes:

#### 1. Restore OAuth CSRF protection.

Entirely encapsulated in [the first commit](https://github.com/code-dot-org/code-dot-org/pull/35752/commits/0bc755e307f4fe107a04656ce2365a6b0bbcde5b), which reverts #35734 

#### 2. Reauthorize Google Classroom by POST

Encapsulated in [this commit](https://github.com/code-dot-org/code-dot-org/pull/35752/commits/5b93902d0e22d674939aed47ccfc77f80cd9216f), we replace the reauth hyperlink in our load failure view (pictured above) with with an awesome orange button that submits a hidden form with an auth token, giving us CSRF-protected reauthorization.

#### 3. Fix the reauthorization flow

There was an independent bug, that's been around a long time now, where we stopped auto-launching the import section flow after reauthorization!  [That's fixed here.](https://github.com/code-dot-org/code-dot-org/pull/35752/commits/6c8b737a5090dd608208779e906b6884893685f1)

#### 4. Clean up RosterDialog's storybook

At some point RosterDialog's prop interface changed significantly, but its stories were not updated to match.  I've fixed up the stories to reflect how we use RosterDialog now, and also done some cleanup on the story file so we generate all stories for all providers and use a decorator to open the dialogs by default in the story view.

## Testing story

I did manual testing locally by forcing a Google Classroom auth failure using the technique described on https://github.com/code-dot-org/code-dot-org/pull/35734. (Note, this requires you to have [OAuth set up on your local machine.](https://docs.google.com/document/d/1QfHFWQ1mN8deGJz6CW9F3UTgReYGADO2P7aRGo0monU/edit))

Here's a screencast of the whole rostering reauth flow.  Note, we get a second load error at the end of the flow because I've hard-coded this for testing.

roster-reauth

# Reviewer Checklist:

- [ ] Tests provide adequate coverage
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

File: apps/src/templates/teacherDashboard/RosterDialog.jsx [hunk 1]
```diff
@@ -11,6 +11,15 @@ import {
   isRosterDialogOpen
 } from './teacherSectionsRedux';
 
+const orangeButtonStyle = {
```
Comment (Hamms) [inline apps/src/templates/teacherDashboard/RosterDialog.jsx]: nit; would it work to call this just "buttonStyle"? I don't love including implementation details like color in the name

**Review by Hamms — APPROVED** [review]
This looks great! Love the tests. One question: would it be worthwhile to drop `authenticityToken` into redux?

**Comment by islemaster** [comment]
> would it be worthwhile to drop authenticityToken into redux?

I puzzled over that a little bit.  How much do we want to generalize here?  I get the impression that we _could_ store one authenticity token in redux... or we could write a helper that just pulls a valid authenticity token from the CSRF meta tags in the document head, [since we are including them everywhere](https://github.com/code-dot-org/code-dot-org/blob/67357c880e971bac6fc7fea901b6408847d70618/dashboard/app/views/layouts/application.html.haml#L27).  I'm going to experiment with that and see if it works.

File: apps/src/templates/teacherDashboard/RosterDialog.jsx (hunk 1 above)
Comment (islemaster) [reply apps/src/templates/teacherDashboard/RosterDialog.jsx]: I went with `ctaButtonStyle` which differentiates this button's role from other buttons on the page without actually saying anything about its style properties.

**Comment by islemaster** [comment]
Pulling the CSRF token from the page's meta tags worked great, and now we actually have less plumbing in this code than we did before.  PTAL!

**Comment by islemaster** [comment]
:facepalm: Forgot to stub those meta tags in tests.  Should be fixed now!

**Review by Hamms — APPROVED** [review]
oh dang, that is _clean_

**Comment by islemaster** [comment]
> oh dang, that is _clean_

:joy: I suppose that depends on who you ask.  I am depending on globals and action-at-a-distance in a way the earlier solution did not.  IMO this is a good use case for those dependencies.

**Comment by islemaster** [comment]
Uh-oh - the latest test failure looks unrelated to this change, but might be a flaky test from the recent AFE work.  Going to restart this test build, and also take a look on another branch.

```
FAIL["test_new-code-account_is_1_if_the_user_was_created_less_than_five_minutes_ago", "Api::V1::AmazonFutureEngineerControllerTest", 26.09636051207781]

test_new-code-account_is_1_if_the_user_was_created_less_than_five_minutes_ago#Api::V1::AmazonFutureEngineerControllerTest (26.10s)
not all expectations were satisfied
unsatisfied expectations:
- expected exactly once, not yet invoked: Services::AFEEnrollment.submit()
```

**Pull request merged**